src/typegate/src/typegate/artifacts/shared.ts

```diff
@@ -70,52 +72,59 @@ class SharedArtifactPersistence implements ArtifactPersistence {
     this.s3.destroy();
   }
 
-  async save(stream: ReadableStream<any>): Promise<string> {
+  async save(stream: ReadableStream<any>, size: number): Promise<string> {
     const hasher = createHash("sha256");
 
-    // TODO compatibility with Node.js streams?
-    // const stream2 = stream.pipeThrough(new HashTransformStream(hasher));
-    //
-    // const tempKey = resolveS3Key(
-    //   `tmp/${Math.random().toString(36).substring(2)}`,
-    // );
-    //
-    // const _ = await this.s3.putObject({
-    //   Bucket: this.s3Bucket,
-    //   Body: stream2,
-    //   Key: tempKey,
-    // });
-    // const hash = hasher.digest("hex");
-    //
-    // await this.s3.copyObject({
-    //   Bucket: this.s3Bucket,
-    //   CopySource: tempKey,
-    //   Key: resolveS3Key(hash),
-    // });
-    //
-    // await this.s3.deleteObject({
-    //   Bucket: this.s3Bucket,
-    //   Key: tempKey,
-    // });
-    //
-    // return hash;
-
-    const tmpFile = await Deno.makeTempFile({ dir: this.dirs.temp });
-    const file = await Deno.open(tmpFile, { write: true, truncate: true });
-    await stream
-      .pipeThrough(new HashTransformStream(hasher))
-      .pipeTo(file.writable);
+    const stream2 = stream.pipeThrough(new HashTransformStream(hasher));
+    
+    const tempKey = resolveS3Key(this.s3Bucket, 
```

Comment:
can you add you a comment that explain how it the save working (e.g. why a tempKey is needed)?